[6.2] Remove ExecRuleHelper

ditsuke opened the pull request:
### Summary of Changes
Replace the ExecRuleHelper class with Task::computeNextExecution().


### Testing Instructions
- Creating new tasks and modifying existing tasks works as expected.
- Triggered tasks computer next execution date as expected from configuration.

### Actual result BEFORE applying this Pull Request

`ExecRuleHelper` exists, for little reason.

### Expected result AFTER applying this Pull Request

`ExecRuleHelper` is safely replaced by `Task::computeNextExecution()`


### Documentation Changes Required
\-

ditsuke committed:
Add ExecRuleHelper replacement to Task driver

- Task::computeNextExecution() replaces ExecRuleHelper::nextExec().
- Replaces ExecRuleHelper() usages.

ditsuke committed:
Remove ExecRuleHelper

ExecRuleHelper is made redundant with the addition of
Task::computeNextExecution(). Given all its usages have been replaced,
this commit safely removes the class.

ditsuke committed:
Fix code-style

chmst commented:
I applied this patch and tried to chage settings of a task:
![settings-task](https://user-images.githubusercontent.com/1035262/147821842-71b7dacf-1799-478b-8957-c2e7dbd5709c.png)
 
When saving this settings: 
![error-task](https://user-images.githubusercontent.com/1035262/147821923-c40d40ef-e306-46c5-9477-86171419ce6f.png)

PHP 8.0, win11 and xampp

richard67 commented:
The removed helper PHP will have to be added to the list of files to be deleted on update in script.php. I will do that when this PR has been merged.

richard67 commented:
@ditsuke I think @chmst indeed found an issue with your changed code, possibly here: https://github.com/joomla/joomla-cms/pull/36439/files#diff-b9f75ae33fad36d7a8968d12a2b78ec972c5e6d2c6fd7ed0024d764191ec48ecR572 .

ditsuke committed:
Fix object resolution in Task constructor

Fixes issue identified with creating/saving tasks. Since Task objects
are also now created by TaskModel::save(), the parameter resolution in
the constructor must accommodate for records from both the database and
form submissions (both cast to object). This commit adds the necessary
check, so we can deal with both possible types.

ditsuke committed:
Merge branch '4.1-dev' of github.com:joomla/joomla-cms into 'replace-exec-rule-helper'

ditsuke commented:
@chmst @richard67 thanks for identifying the issue! The latest commit should solve it. Please test.

chmst commented:
You fixed the code error, but I don't understand why the Task Params are twice. First in tab "New Task" and in tab "Advanced"?

ditsuke commented:
> You fixed the code error, but I don't understand why the Task Params are twice. First in tab "New Task" and in tab "Advanced"?

@chmst I addressed that in https://github.com/joomla/joomla-cms/pull/36515 (needs test). We need the routine parameter fieldset to have a `showFront="true"` attribute at the moment. Although I don't think that's ideal, this PR should work for now. Let's try to work on removing this requirement in the future!

bembelimen committed:
Merge branch '4.1-dev' into replace-exec-rule-helper

ditsuke committed:
Merge branch '4.1-dev' of github.com:joomla/joomla-cms into 'replace-exec-rule-helper'

ditsuke committed:
Merge branch '4.1-dev' into replace-exec-rule-helper

ditsuke commented:
Could we please get a couple tests on this 😃

ditsuke committed:
Merge remote-tracking branch 'jcms/4.1-dev' into replace-exec-rule-helper

Conflicts:
	administrator/components/com_scheduler/src/Task/Task.php

chmst commented:
@ditsuke this PR is still open. could you please resolve the conflicts so that we can re-test?

HLeithner commented:
This pull request has been automatically rebased to 5.1-dev.

HLeithner commented:
This pull request has been automatically rebased to 5.2-dev.

cybersalt commented:
I have tested this item :red_circle: unsuccessfully on 4f91c6a8744af9eca1750c531e2edd824e52e798<br /><br />I opened an existing task, changed a setting, saved and closed and got a 500 error:

An error has occurred.
0 Class "Joomla\Component\Scheduler\Administrator\Model\ExecRuleHelper" not found
Call Stack
#	Function	Location
1	()	JROOT/administrator/components/com_scheduler/src/Model/TaskModel.php:571
2	Joomla\Component\Scheduler\Administrator\Model\TaskModel->save()	JROOT/libraries/src/MVC/Controller/FormController.php:689
3	Joomla\CMS\MVC\Controller\FormController->save()	JROOT/libraries/src/MVC/Controller/BaseController.php:730
4	Joomla\CMS\MVC\Controller\BaseController->execute()	JROOT/libraries/src/Dispatcher/ComponentDispatcher.php:143
5	Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch()	JROOT/libraries/src/Component/ComponentHelper.php:361
6	Joomla\CMS\Component\ComponentHelper::renderComponent()	JROOT/libraries/src/Application/AdministratorApplication.php:150
7	Joomla\CMS\Application\AdministratorApplication->dispatch()	JROOT/libraries/src/Application/AdministratorApplication.php:195
8	Joomla\CMS\Application\AdministratorApplication->doExecute()	JROOT/libraries/src/Application/CMSApplication.php:306
9	Joomla\CMS\Application\CMSApplication->execute()	JROOT/administrator/includes/app.php:58
10	require_once()	JROOT/administrator/index.php:32<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36439">issues.joomla.org/tracker/joomla-cms/36439</a>.</sub>

fgsw commented:
@cybersalt same as [this comment](https://github.com/joomla/joomla-cms/pull/35101#issuecomment-2324049033)

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

Hackwar committed:
Merge branch '6.0-dev' into replace-exec-rule-helper

# Conflicts:
#	administrator/components/com_scheduler/src/Helper/ExecRuleHelper.php
#	administrator/components/com_scheduler/src/Model/TaskModel.php
#	administrator/components/com_scheduler/src/Task/Task.php

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle

QuyTon commented on administrator/components/com_scheduler/src/Task/Task.php:
```diff
@@ -559,4 +563,41 @@ public static function isValidId(string $id): bool
+     *
+     * @return ?Date|string  Next due execution.
+     *
+     * @since  4.1.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

HLeithner commented:
This pull request has been automatically rebased to 6.1-dev.

Hackwar committed:
Update administrator/components/com_scheduler/src/Task/Task.php

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

Hackwar committed:
Merge branch '6.2-dev' into replace-exec-rule-helper
